x/genutil/types/genesis.go

```diff
@@ -0,0 +1,157 @@
+package types
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"time"
+
+	cmtjson "github.com/cometbft/cometbft/libs/json"
+	cmtproto "github.com/cometbft/cometbft/proto/tendermint/types"
+	cmttypes "github.com/cometbft/cometbft/types"
+	cmttime "github.com/cometbft/cometbft/types/time"
+
+	"github.com/cosmos/cosmos-sdk/version"
+)
+
+const (
+	// MaxChainIDLen is the maximum length of a chain ID.
+	MaxChainIDLen = cmttypes.MaxChainIDLen
+)
+
+// AppGenesisOnly defines the app's genesis.
+type AppGenesis struct {
+	AppName       string          `json:"app_name"`
+	AppVersion    string          `json:"app_version"`
+	GenesisTime   time.Time       `json:"genesis_time"`
+	ChainID       string          `json:"chain_id"`
+	InitialHeight int64           `json:"initial_height"`
+	AppHash       []byte          `json:"app_hash"`
+	AppState      json.RawMessage `json:"app_state,omitempty"`
+
+	// TODO eventually abstract from CometBFT types
+	Validators      []cmttypes.GenesisValidator `json:"validators,omitempty"`
+	ConsensusParams *cmtproto.ConsensusParams   `json:"consensus_params,omitempty"`
+}
+
+// NewAppGenesisWithVersion returns a new AppGenesis with the app name and app version already.
+func NewAppGenesisWithVersion(chainID string, appState json.RawMessage) *AppGenesis {
+	return &AppGenesis{
+		AppName:    version.AppName,
+		AppVersion: version.Version,
+		ChainID:    chainID,
+		AppState:   appState,
+		Validators: nil,
+	}
+}
+
+// ToCometBFTGenesisDoc converts the AppGenesis to a CometBFT GenesisDoc.
+func (ag *AppGenesis) ToCometBFTGenesisDoc() (*cmttypes.GenesisDoc, error) {
+	var consensusParams *cmttypes.ConsensusParams
+	if ag.ConsensusParams != nil {
+		consensusParams = &cmttypes.ConsensusParams{
+			Block: cmttypes.BlockParams{
+				MaxBytes: ag.ConsensusParams.Block.MaxBytes,
+				MaxGas:   ag.ConsensusParams.Block.MaxGas,
+			},
+			Evidence: cmttypes.EvidenceParams{
+				MaxAgeNumBlocks: ag.ConsensusParams.Evidence.MaxAgeNumBlocks,
+				MaxAgeDuration:  ag.ConsensusParams.Evidence.MaxAgeDuration,
+				MaxBytes:        ag.ConsensusParams.Evidence.MaxBytes,
+			},
+			Validator: cmttypes.ValidatorParams{
+				PubKeyTypes: ag.ConsensusParams.Validator.PubKeyTypes,
+			},
+		}
+	}
+
+	return &cmttypes.GenesisDoc{
+		ChainID:         ag.ChainID,
+		InitialHeight:   ag.InitialHeight,
+		AppHash:         ag.AppHash,
+		AppState:        ag.AppState,
+		ConsensusParams: consensusParams,
+		Validators:      ag.Validators,
```

Comment:
can we add a regression test for this function. I know tmjson has a different style encoding for keys/interfaces that may affect this.

Reply:
Added in [eb83d6a](https://github.com/cosmos/cosmos-sdk/pull/15031/commits/eb83d6ae94693e2e07a0941b0a4888112aa1dfc8)